ui/src/components/Exchange/MarketPairs.tsx

```diff
@@ -32,7 +32,17 @@ const MarketPairs: React.FC<Props> = ({ sideNav, onLogout }) => {
     ));
 
     const header = ['Pair', 'Current Price', 'Change', 'Volume']
-    const rows = exchangeContract?.contractData.tokenPairs.map(pair => {
+    const collateralizedRows = exchangeContract?.contractData.tokenPairs.map(pair => {
+        const [ base, quote ] = unwrapDamlTuple(pair);
+        const pairLabel = <>{base.label} <ExchangeIcon/> {quote.label}</>;
+        return [pairLabel, '-', '-', '-'];
+    }) || [];
+
+    const clearedRows = exchangeContract?.contractData.clearedMarkets.map(listing => {
+        const pair = unwrapDamlTuple(listing)[0];
+        if (typeof pair === 'string') {
+            throw new Error('blah')
```

Comment:
Bob Loblaw's Law Blog

Reply:
I was very disappointed that I had to do this, typescript can't infer the fact that this is a cleared market so the first element of the outer tuple here will always be another tuple (the Id pair)

This error would (should) never get thrown, since it's just here to satisfy the compiler... but I will change the error message though 😆

Reply:
Weird that it's totally fine with you assuming it's a tuple only provided it isn't a string...what if it's literally any other type??

Reply:
That's the cool thing, the compiler knows at least it can't be any other type. On the Daml side, `clearedMarkets` contains items of type `((Id, Id), Party)`

and with the typescript codegen, this translates to `Tuple2<Tuple2<Id, Id>, String>` 

Then when calling `unwrapDamlTuple(listing)`, the type is converted to `(string | Tuple2<Id,Id>)[]` - an array of items that could either be strings or tuples, including the first item (although we know in advance it maps to the tuple, this may not be true just by the type) 

So in unwrapping, we actually lose some type information that we had before! Interesting, isn't it?

Reply:
Ohhhhh I see! Super interesting, thank you for the explanation